File: Week4/Solver.java
import java.util.Stack;
import java.util.LinkedList;

public class Solver 
{
    private boolean isSolvable;
    private Node finishNode;

    private class Node implements Comparable<Node>
    {
        private Node parent;
        private int priority;
        public int moves;
        public Board board;

        public Node(Node parentIn, Board boardIn)
        {
            parent = parentIn;
            if (parent == null)
                moves = 0;
            else
                moves = parent.moves + 1;

            board = boardIn;
            priority = board.manhattan();
        }

        public int priority()
        { return priority + moves; }

        public int compareTo(Node that)
        {
            return Integer.compare(priority(), that.priority());
        }
    }

    public Solver(Board initial)            // find a solution to the initial board (using the A* algorithm)
    {
        Node currentNode;
        MinPQ<Node> queue = new MinPQ<Node>();
        queue.insert(new Node(null, initial));
        MinPQ<Node> queueTwins = new MinPQ<Node>();
        queueTwins.insert(new Node(null, initial.twin()));

        int steps = 0;

        while (true)
        {
            currentNode = queue.delMin();

            //StdOut.println("Step "); StdOut.print(steps++);
            //StdOut.println("Get out nod with priority ");
            //StdOut.print(currentNode.priority());
            //StdOut.println(currentNode.board);

            if (currentNode.board.isGoal())
            {
                isSolvable = true;
                finishNode = currentNode;
                return;
            }
            for (Board newBoard : currentNode.board.neighbors())
            {

                if (currentNode.parent != null && newBoard.equals(currentNode.parent.board)) 
                {
                    //StdOut.println("Skipped board");
                    //StdOut.println(currentNode.board);
                    continue;
                }
                queue.insert(new Node(currentNode, newBoard));
            }

            currentNode = queueTwins.delMin();
            if (currentNode.board.isGoal())
            {
                isSolvable = false;
                finishNode = currentNode;
                return;
            }
            for (Board newBoard : currentNode.board.neighbors())
            {
                if (newBoard.equals(currentNode.parent)) continue;
                queueTwins.insert(new Node(currentNode, newBoard));
            }

        }

    }

    public boolean isSolvable()             // is the initial board solvable?
    {
        return isSolvable;
    }

    public int moves()                      // min number of moves to solve initial board; -1 if no solution
    {
        if (isSolvable) return finishNode.moves;
        return -1;
    }

    public Iterable<Board> solution()       // sequence of boards in a shortest solution; null if no solution
    {
        if (!isSolvable) return null;

        LinkedList<Board> result = new LinkedList<Board>();
        Node currentNode = finishNode;
        while (currentNode != null)
        {
            result.addFirst(currentNode.board);
            currentNode = currentNode.parent;
        }

        return result;

    }

    public static void main(String[] args)  // solve a slider puzzle (given below)
    {
        In in = new In(args[0]);
        int N = in.readInt();
        int[][] blocks = new int[N][N];
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                blocks[i][j] = in.readInt();
        Board initial = new Board(blocks);


        // solve the puzzle
        Solver solver = new Solver(initial);

        // print solution to standard output
        if (!solver.isSolvable())
            StdOut.println("No solution possible");
        else 
        {
            StdOut.println("Minimum number of moves = " + solver.moves());
            for (Board board : solver.solution())
                StdOut.println(board);
        }
    }

}

File: Week4Cpp/Solver.cpp
//
//  Solver.cpp
//  Week4
//
//  Created by Вячеслав Ксенз on 23.03.14.
//  Copyright (c) 2014 Вячеслав Ксенз. All rights reserved.
//

#include <vector>
#include <queue>
#include "Board.cpp"

class Node
{
private:
    int priorityCache;
public:
    Node* parent;
    int moves;
    Board* thisBoard;
    
    Node(Node* parentIn, Board *boardIn)
    {
        parent = parentIn;
        if (parent == NULL)
            moves = 0;
        else
            moves = parent->moves + 1;
        thisBoard = boardIn;
        priorityCache = thisBoard->manhattan() + moves;
    }
    
    int priority()
    {
        return priorityCache;
    }
    
    ~Node()
    {
        delete thisBoard;
    }
};

class NodeCompare
{
public:
    bool operator()(Node* x, Node* y)
    {
        return x->priority() > y->priority();
    }
};

class Solver
{
private:
    
    bool isSolvable;
    Node* finishNode;
    
    
public:
    
    Solver(Board *initial)
    {
        Node* currentNode;
        std::vector<Board *> neighbors;
        std::priority_queue<Node *, std::vector<Node *>, NodeCompare> pq_main;
        std::priority_queue<Node *, std::vector<Node *>, NodeCompare> pq_twins;
        pq_main.push(new Node(NULL,initial));
        pq_twins.push(new Node(NULL,initial->twin()));
        int step = 0;
        while (1)
        {
            step++;
            
            // Get the min Node
            currentNode = pq_main.top();
            pq_main.pop();
            
            // If it the end Node then we finished
            if (currentNode->thisBoard->isGoal())
            {
                isSolvable = 1;
                finishNode = currentNode;
                return;
            }
            
            // Get all the neighbors and push them in the queue
            neighbors = currentNode->thisBoard->neighbors();
            for (auto it: neighbors)
            {
                if (currentNode->parent != NULL && it->isEqual(currentNode->parent->thisBoard))
                    continue;
                
                pq_main.push(new Node(currentNode, it));
            }

            // Repeat all the same, but for the twins queue
            currentNode = pq_twins.top();
            pq_twins.pop();
            // If it the end Node then we finished
            if (currentNode->thisBoard->isGoal())
            {
                isSolvable = 1;
                finishNode = currentNode;
                return;
            }
            
            // Get all the neighbors and push them in the queue
            neighbors = currentNode->thisBoard->neighbors();
            for (auto it: neighbors)
            {
        
                if (currentNode->parent != NULL && it->isEqual(currentNode->parent->thisBoard))
                    continue;
                
                pq_twins.push(new Node(currentNode, it));
            }

            
        }
        
    }
    
    bool isSolvavle()
    {
        return isSolvable;
    }
    
    int moves()
    {
        if (isSolvable) return finishNode->moves;
        return -1;
    }
    
    std::vector<Node *> solution()
    {
        std::vector<Node *> result;
        if (!isSolvable) return result;
        
        Node *currentNode = finishNode;
        while (currentNode != NULL)
        {
            result.push_back(currentNode);
            currentNode = currentNode->parent;
        }
        
        return result;
        
    }
    
    ~Solver()
    {
        //delete finishNode;
    }
    
};